<center>
    
## <font color='maroon'>ASTR 20500</font>
## <font color='maroon'>homework assignment 1</font>
    
#### <font color='black'>distributed: Monday, October 4</font>
#### <font color='black'>due: Friday, October 8 (9pm)</font>
    
    
#### 30 points total (+15 potential extra-credit points)
</center>


Information you need to carry out exercises in this assignment is available in the first two distributed notebooks: [01_python_intro](https://drive.google.com/file/d/1NDPRhf5zLu0I7Bcs3JtutqhpzE_5i684/view?usp=sharing) and [02_probability](https://drive.google.com/file/d/18r-zCP4W67VW5HAOElvmpXkTB2Q_kX88/view?usp=sharing) or is provided here. 

Specific background info useful for this homework can be found in [hw01_background_info]() notebook. Make sure you review it. 

Although there are 5 exercises listed, default is 4 exercises in this assignment. You can **_choose either exercise 3 or 4_** for the third exercise. Exercise 3 is somewhat simpler conceptually and requires a bit less knowledge of programming and is recommended to students new to both Python/programming and statistics. Exercise 4 is recommended for students who have already some knowledge of Python and are a bit comfortable with statistics. If you do both exercises, exercise 3 will be counted as extra-credit.   

### <font color='blue'>Exercise 1: practicing programming with loops (5 points)</font> 

**Task 1a. (3 points)** Use this type of <tt>for</tt> loop to compute factorial $n!=n\,(n-1)\,(n-2)\,\ldots 2\cdot\,1$ for a specified $n$ and print out result. 

In [1]:
n = int(input("Pick a number: "))

if n<0:
  print("Invalid input! Please rerun the program and enter a non-negative integer.")
elif n==0:
  print(1)
else:
  for i in range(1,n):
    if i<n:
      n *= i
  print(n)

Pick a number: 4
24


**Task 1b. (2 points)** Use the code you developed in 1a to implement a function for computing factorial of a specified integer $n$ with this header:

In [2]:
def myfactorial(n):
  if n<0:
    print("Invalid input! Please rerun the program and enter a non-negative integer.")
  elif n==0:
    return 1
  else:
    for i in range(1,n):
      if i<n:
        n *= i
    return n

myfactorial(int(input("Pick a number: ")))

Pick a number: 4


24

At the beginning of the function include a check that $n\geq 0$ end ensure that input <tt>n</tt> is of type <tt>int</tt>. If $n<0$ or the type of input is different from <tt>int</tt> the function should print a meaningful error message and return nothing (<tt>None</tt>).   


### <font color='blue'>Exercise 2:  (5 points)</font> 

Implement a function computing the binomial probability for input probability of an outcome, $p$, the number of outcomes $n$, and the number of experiments $N$: 

$$P(n,N)=\frac{N!}{n!(N-n)!}\, p^n(1-p)^{N-n}.$$ 

The function should use the myfactorial function implemented in exercise 1. and have the following format: 

In [3]:
def pbinomial(p, N, n):
    P=myfactorial(N)/(myfactorial(n)*myfactorial(N-n))*p**n*(1-p)**(N-n)
    return P
p=float(input("Enter the probability: "))
N=int(input("Enter the number of experiments: "))
n=int(input("Enter the number of outcomes: "))
print("The probability of this scenario occuring is: " + str(pbinomial(p, N, n)))

Enter the probability: .5
Enter the number of experiments: 10
Enter the number of outcomes: 5
The probability of this scenario occuring is: 0.24609375


Test your function using a simple case where you can compute the answer independently by hand and compare it to the output of your function. 

### <font color='blue'>Exercise 3:  (10 points)</font> 

**Background info.** To accurately measure flux (brightness) of an astronomical object observing conditions must be sufficiently good. For example, there should be nearly no clouds, Moon light should be minimal, etc. Observing nights during which such conditions are met are called "*photometric*" *nights* in the astronomical jargon. Non-photometric  nights can be used for other types of observations. 

At the [Las Campanas Observatory](http://www.lco.cl/las-campanas-observatory/) located high in the southern reaches of Chile’s Atacama Desert, where many UChicago astronomers do their observations, the fraction of photometric nights varies from 0.4 to 0.8 during the year (for example, Figure 5-8 [here](https://www.gmto.org/Resources/GMT-ID-01466-Chapter_5_Site_Evaluation.pdf)). 

**Task 3a. (7 points)** Suppose you are involved in preparation for an observational campaign at the Las Campanas Observatory. You figured out that to achieve your goals, you will need at least 4 photometric nights. You and your collaborators receive 7 observing nights. What is the probability that you will complete the observations you need in this observing run and will not need to apply for more observing time in the worst case scenario? 

Present your arguments and complete calculations in the text and code cells below. 
Also, compute the same estimate for the best case scenario. The difference between these can be viewed as a rough estimate of uncertainty. 


In the worst case scenario, the fraction of photometric nights is $p=0.4$, so if we treat this as our probability p and utilize the binomial probability function from Exercise 2, we can calculate the probability we will get four photometric nights out of our seven:

In [4]:
print("The probability of getting the worst case scenario of photometric nights is", pbinomial(0.4,7,4))

The probability of getting the worst case scenario of photometric nights is 0.193536


As such, in the worst case scenario, the probability that our observational campaign at Las Campanas Observatory succeeds at getting four photometric nights out of seven and not having to reapply is not terrible at $\boxed{P\approx 0.19}$ . On the other hand, the best case scenario will be when the fraction of photometric nights we get is $p=0.8$, so doing the same thing we did up above but with $p=0.8$ instead of $p=0.4$, but adding up the binomial probabilities of having 4 to 7 photometric nights since all we need is *at least* four photometric nights, we find:

In [5]:
P=0
for n in range(4,8):
  P += pbinomial(0.8,7,n)
print("The probability of getting the best case scenario of photometric nights is", P)

The probability of getting the best case scenario of photometric nights is 0.9666560000000001


As such, in the best case scenario, the probability that our observational campaign at Las Campanas Observatory succeeds at getting \emph{at least} four photometric nights out of seven and not having to reapply is pretty high at $\boxed{P\approx 0.97}$ .

**Task 3b. (3 points)** For a given assumed fraction of photometric nights $p=0.4$ and keeping $N=7$ (the total number of nights), compute and print binomial probability for different values of $n$ from 0 to 7. Discuss how the binomial probability changes with $n$ and what can you notice about its distribution?

In [6]:
print("n","Probability")
for n in range(0,8):
  print(n,pbinomial(0.4, 7, n))

n Probability
0 0.027993599999999993
1 0.13063679999999997
2 0.2612736
3 0.29030400000000006
4 0.193536
5 0.07741440000000002
6 0.017203200000000005
7 0.0016384000000000006


Observing the probability distribution above for $n\in[0,7]$, we can see that the binomial probability increases as we get closer to the middle and peak value of $n$ at $n=3$, and kind of fades (gets smaller) as we get to the edges. This is basically what we expected since we are conducting a binomial distribution, which has the larger values in the middle of the distribution and the smaller values at the edges (think normalized bell curve).

### <font color='blue'>Exercise 4:  (10 points + 5 extra-credit)</font> 

**Background info.** [Supernovae](https://www.youtube.com/watch?v=B94yHU-e5_c) are stars that brighten ten billion times or more  compared to normal stars for a short period of time (days). Supernovae events are the last stage of evolution of massive stars or binary [white dwarfs](https://imagine.gsfc.nasa.gov/science/objects/dwarfs2.html). 
Since the year 1000 (when more or less reliable records exist) 5 bright supernovae occurred that were [observed in the sky without telescope](https://astronomy.com/news/2020/11/7-naked-eye-supernovae-throughout-human-history) in 1006, 1054, 1181, 1572, and 1604. 

**Task 4a. (5 points)** Assuming that probability of such bright supernovae to occur in a given century is constant, estimate this probability using *two* different methods. 

*Hint:* think about the value of probability that would maximize the chance of having 5 supernovae events in 10 centuries. 

Method 1 of computing the probability of five bright supernovae occuring in a given millenia assuming constant probability per century:

P(Supernovae/century)= Num. of Supernovae / Num. of Centuries of available data: = $$\text{P(Supernovae/Century)}=\frac{5}{10}=\boxed{0.5}$$

Method 2 of computing the probability involves utilizing a for loop to test every possible probability (to two significant figures) and find the maximum for a binomial probability of 5 outcomes in 10 experiments.

In [12]:
p=0.01
P=0
print("Probability","Binomial Probability")
while p<1:
  P=pbinomial(p, 10, 5)
  print(p,P)
  p += 0.01

Probability Binomial Probability
0.01 2.396494925748e-08
0.02 7.2892173053952e-07
0.03 5.258543679776519e-06
0.04 2.104052938702848e-05
0.05 6.0935248828124994e-05
0.060000000000000005 0.0001438123094901965
0.07 0.0002946493474551925
0.08 0.0005442389404758838
0.09 0.0009285815482274495
0.09999999999999999 0.0014880347999999995
0.10999999999999999 0.0022662822629686643
0.11999999999999998 0.003309178025693673
0.12999999999999998 0.0046635170125185164
0.13999999999999999 0.006375774895905296
0.15 0.008490855786328123
0.16 0.011050880543993365
0.17 0.014094043564225089
0.18000000000000002 0.01765356122970059
0.19000000000000003 0.02175673088869989
0.20000000000000004 0.02642411520000001
0.21000000000000005 0.03166886297285501
0.22000000000000006 0.03749617421551155
0.23000000000000007 0.043902913978773016
0.24000000000000007 0.05087737673309754
0.25000000000000006 0.05839920043945319
0.26000000000000006 0.06643942715651133
0.2700000000000001 0.0749607049605932
0.2800000000000001 0.083917

We can see that the maximum binomial probability is $P=0.24609375$, which occurs at $p=0.50$, implying that the probability of a bright supernovae occuring in a century is $\boxed{p=0.50}$ .

**Task 4b. (5 points)** Present at least two different arguments backed by calculations and numbers that the number of recorded bright supernovae events and the years in which they occurred are *statistically* consistent with the constant probability estimated in 1a. 

One possible way we can show that the number of recorded bright supernovae events and the years in which they occured are statistically consistent is by calculating the probability that we would have four centuries within a millenia that do not have supernovae and see if that's somewhat realistic. We know from part (a) that there is a p=0.5 chance of having (and therefore not having) a supernovae in a given century, so we can find the probability that we wouldn't have a supernovae in four consecutive centuries:

$$P=(0.5)^4=0.0625\approx \boxed{6.3\%}$$

which isn't the most unrealistic of answers (its larger than 1% and even 5% so we're good!). Considering the Earth has been around for nearly 4.6 billion years now, and with a probability of having a bright supernovae in a given century being p=0.5, it would make sense that we would have at least four centuries in a row without one at some point in the Earth's history. Let's even just consider the duration that humans have been on Earth for -- about 6 million years. That's 60,000 centuries, implying that in that duration, we would have had $60000\cdot 0.0625=3750$ blocks of four consecutive centuries that did not have supernovae. That's a lot. So it is definitely not unrealistic that we had one such scenario in the last millenia.

Another possible way we can show that the number of record bright supernovae events and the years in which they occured are statistically consistent is by checking how realistic it is to have three supernovae within the first half millenia (five centuries) of the years 1000-2000 BCE. Since we had three supernovae in 500 years, or 5 centuries, this implies a probability of having a supernovae per century (according to the calculations we did in part(a)) of:

$$\text{P(Supernovae/Century}=\frac{3}{5}=\boxed{0.60}$$

Since our sample size is smaller here, it would make sense that the probability isn't the same as what we calculated in part (a), but a probability of $P=0.60$ supernovae/century is pretty darn close to our expected $P=0.50$. In fact, that's only about $20\%$ error. As such, we find that our data is statistically consistent even on smaller scales. Having more data points would be more helpful for finding the true value for the probability of having a supernovae in a century (as extrapolating wouldn't be a very good idea here), but otherwise we have completed our goal in finding two methods that show that our data for the number of recorded bright supernovae events and the years in which they occured are statistically consistent.

**Task 4c. (optional extra-credit 5 points)** Devise a way to estimate uncertainty of your estimate of probability in 1a and estimate it.  

*Hint:* the uncertainties are often estimated using a width of probability distribution. One way to define the width of a distribution is by looking at the so-called ["Full Width at Half Maximum" (FWHM)](https://en.wikipedia.org/wiki/Full_width_at_half_maximum) - the distance between values of parameter for which probability is twice smaller than at the maximum. 

We can use the Full Width at Half Maximum (FWHM) to find out how narrow our binomial distribution is -- i.e. quantify our spread of data points and see how realistic/unrealistic they are. If we look at the printout table for binomial probability at each probability point from part (a), we see that our maximum occurs at $p=0.5$, specifically being $P\approx 0.246$. Half of our maximum $P$ is $P_{\text{HM}}=\frac{0.246}{2}\approx 0.123$, which is very close to the binomial probability outputs for $p_1=0.32$ and $p_2=0.68$. Finding the difference between $p_1$ and $p_2$ will give us the Full Width at Half Maximum FWHM:

$$\text{FWHM}=p_1-p_2=0.68-0.32=\boxed{0.36}$$

A FWHM of 0.36 is comparatively small, indicating that the distribution of our data is narrower rather than wider (a typical FWHM could be around 0.5, so 0.36 is definitely smaller), indicating a lower uncertainty in our data.

### <font color='blue'>Exercise 5:  (10 points)</font> 

Tests for COVID-19 (or any other disease) are characterized by their probability to produce positive result when disease is present, $P(+\vert C)$ (where $C$ stands for COVID presence),  which is called *test sensitivity*. 

Another important characteristic of a test is *specificity* defined as the probability that a test result will be negative when disease is not present, $P(-\vert N)$ (where $N$ stands for "No disease present"). Specificity is a complement probability of the probability to get positive test result when disease is not present, $P(+\vert N)=1-P(-\vert N)$, which is called the *false positive rate*.

Probability for a random member of a community to have COVID, P(C), can be conservatively approximated by the current positivity rate (the fraction of people in the community who test positive). 

**Task.** 
Suppose you know when you are taking a COVID test that its sensitivity is $P(+\vert C)=0.97$, specificity is $P(-\vert N)=0.995$ and the current COVID positivity rate on campus is 1\% so $P(C)= 0.01$. 

If you test positive with this test, compute the probability that you *actually* have COVID-19? 

How does the result change if the test specificity is $P(-\vert N)=0.97$ instead? What can you conclude about importance of knowing specificity of such tests?

*Hint:* use the Bayes theorem and the "law of total probability", which were discussed on Monday and are described in [02_probability](https://drive.google.com/file/d/18r-zCP4W67VW5HAOElvmpXkTB2Q_kX88/view?usp=sharing) notebook. 

*Note:* that this exercise does not require much coding, but it requires a derivation of equation for the corresponding probability and calculation of that probability. You should present your argument, equations, and calculations of the probabilities.

In order to find the probability that you have COVID given that you tested positive $P(C\vert +)$, it will suffice to utilize Bayes' Theorem to do so. From Bayes' Theorem, we have that:

$$P(C\vert+)=P(+\vert C)\cdot\frac{P(C)}{P(+)}$$

We are given both the probability of having COVID on campus, $P(C)=0.01$ and the probability of testing positive given that you have COVID, $P(+\vert C)=0.97$, so all we need to find is the probability of testing positive $P(+)$. There are two scenarios where you can test positive -- the first is if you have COVID and test positive $P(+\vert C)$, and the second is that you don't have COVID and test positive $P(+\vert N)$. Using a simple probability tree, we find that: 

$$P(+\text{ and }C)=P(+\vert C)\cdot P(C)\\ P(+\text{ and }N)=P(+\vert N)\cdot P(N)$$

Since the probability of testing positive $P(+)$ is the sum total of probabilities of testing positive given you have COVID and not having COVID (as those are the only two scenarios), then we can simply add the two probabilities $P(+\vert C)$ and $P(+\vert N)$ to get $P(+)$, which gives us: 

$$P(+)=P(+\text{ and }C)+P(+\text{ and } N)=P(+\vert C)\cdot P(C)+P(+\vert N)\cdot P(N)$$

Plugging this back into our equation for $P(C\vert +)$ from earlier to find the probability of having COVID given that you tested positive, we find:

$$P(C\vert +)=P(+\vert C)\cdot\frac{P(C)}{P(+\vert C)\cdot P(C)+P(+\vert N)\cdot P(N)}\\=\frac{0.97\cdot 0.01}{0.97\cdot 0.01+0.005\cdot 0.99}=0.6610\approx \boxed{0.66}$$

If the test specitivity is $P(-\vert N)=0.97$ instead of $P(-\vert N)=0.995$, then all we need to do is change that value in our equation for $P(C\vert +)$. Doing so, we find a new probability of having COVID given that you tested positive of:

$$P(C\vert +)=\frac{0.97\cdot 0.01}{0.97\cdot 0.01+0.03\cdot 0.99}=0.2462\approx \boxed{0.25}$$